From 42200147fa2202c13a9d3a1a38d8f254776bf310 Mon Sep 17 00:00:00 2001
From: Ian Dannapel <ian.dannapel@iris-sensing.com>
Date: Fri, 13 May 2022 15:49:59 +0200
Subject: [PATCH] imx8-mipi-cs2-sam: Add Raw12 and Raw14 formats

- fix MEDIA MIX gasket0 dumps
Somebody forgot that the gasket regmap for the media mix already starts
at 0x60. Use an additional counter but keep the offsets within the
print, since this is at least more clear to me.

- Enable summary debug output
- allow format propagation to the subdev
- Hack: Set sink pad to force subdev format change

Signed-off-by: Rouven Czerwinski <r.czerwinski@pengutronix.de>
Signed-off-by: Erik Schumacher <erik.schumacher@iris-sensing.com>
Signed-off-by: Ian Dannapel <ian.dannapel@iris-sensing.com>
Signed-off-by: Jasper Orschulko <jasper@fancydomain.eu>
---
 drivers/staging/media/imx/imx8-mipi-csi2-sam.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/drivers/staging/media/imx/imx8-mipi-csi2-sam.c b/drivers/staging/media/imx/imx8-mipi-csi2-sam.c
index 1143ce3ddd4b0..2e630e8d1dafd 100644
--- a/drivers/staging/media/imx/imx8-mipi-csi2-sam.c
+++ b/drivers/staging/media/imx/imx8-mipi-csi2-sam.c
@@ -447,7 +447,7 @@ struct csi_state {
 	u32 val;
 };
 
-static int debug;
+static int debug = 1;
 module_param(debug, int, 0644);
 MODULE_PARM_DESC(debug, "Debug level (0-2)");
 
@@ -608,13 +608,15 @@ static void dump_gasket_regs(struct csi_state *state, const char *label)
 		{ 0x64, "GPR_GASKET_0_HSIZE" },
 		{ 0x68, "GPR_GASKET_0_VSIZE" },
 	};
-	u32 i, cfg;
+	u32 i, cfg, p;
 
 	v4l2_dbg(2, debug, &state->sd, "--- %s ---\n", label);
 
+	p = 0x0;
 	for (i = 0; i < ARRAY_SIZE(registers); i++) {
-		regmap_read(state->gasket, registers[i].offset, &cfg);
+		regmap_read(state->gasket, p, &cfg);
 		v4l2_dbg(2, debug, &state->sd, "%20s[%x]: 0x%.8x\n", registers[i].name, registers[i].offset, cfg);
+		p += 0x4;
 	}
 }
 
@@ -1185,7 +1187,11 @@ static int mipi_csis_set_fmt(struct v4l2_subdev *mipi_sd,
 		return -EINVAL;
 	}
 
-	format->pad = source_pad->index;
+	/*
+	 * Set the sink instead of the source pad to force the format on the
+	 * tc358746 subdevice: pad 0 is its input and pad 1 its output.
+	 */
+	format->pad = 0;
 	ret = v4l2_subdev_call(sen_sd, pad, set_fmt, NULL, format);
 	if (ret < 0) {
 		v4l2_err(&state->sd, "%s, set sensor format fail\n", __func__);
-- 
2.55.0

